testing/trino-testing/src/main/java/io/trino/testing/TestingConnectorBehavior.java

```diff
@@ -13,16 +13,52 @@
  */
 package io.trino.testing;
 
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
 public enum TestingConnectorBehavior
 {
     SUPPORTS_PREDICATE_PUSHDOWN,
-    SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_EQUALITY,
-    SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY,
+    SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_EQUALITY(SUPPORTS_PREDICATE_PUSHDOWN),
+    SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY(SUPPORTS_PREDICATE_PUSHDOWN),
     SUPPORTS_LIMIT_PUSHDOWN,
     SUPPORTS_TOPN_PUSHDOWN,
     SUPPORTS_AGGREGATION_PUSHDOWN,
-    SUPPORTS_JOIN_PUSHDOWN,
-    SUPPORTS_JOIN_PUSHDOWN_WITH_FULL_JOIN,
-    SUPPORTS_JOIN_PUSHDOWN_WITH_DISTINCT_FROM,
+    SUPPORTS_JOIN_PUSHDOWN(
+            // Currently no connector supports Join pushdown by default. JDBC connectors may support Join pushdown and BaseJdbcConnectorTest
+            // verifies truthfulness of SUPPORTS_JOIN_PUSHDOWN declaration, so it is a safe default.
+            false),
+    SUPPORTS_JOIN_PUSHDOWN_WITH_FULL_JOIN(SUPPORTS_JOIN_PUSHDOWN),
+    SUPPORTS_JOIN_PUSHDOWN_WITH_DISTINCT_FROM(SUPPORTS_JOIN_PUSHDOWN),
     /**/;
+
+    private final Function<Function<TestingConnectorBehavior, Boolean>, Boolean> defaultBehavior;
```

Comment:
I'd probably move the fallback-checking behavior entirely to `TestingConnectorBehavior` as well.

```java
enum TestingConnectorBehavior {
    ...;
    Optional<Boolean> defaultValue;
    Optional<TestingConnectorBehavior> fallback;

    boolean isSupported(Function<TestingConnectorBehavior, Optional<Boolean>> supportChecker)
    {
        return supportChecker.apply(this)
                .or(() -> this.defaultValue)
                .or(() -> this.fallback.map(fallback -> fallback.isSupported(supportChecker)))
                .orElseThrow(() -> new IllegalArgumentException("TestingConnectorBehavior has no default or fallback"));
    }
}
```

Reply:
I agree it would be better from perspective of API between TestingConnectorBehavior and the SmokeTests class.

This design is meant to be optimized for subclass simplicity (eg simple logic, no optionals), while being forward compatibile.

Reply:
I am changing from `Function` to `Predicate`
and changing `getBehavior` to `hasBehavior` (in test classes)
and changing `getDefaultBehavior` to `hasBehaviorByDefault` (in the enum)